Support generation of last branch events with perf inject

sebpop opened the pull request.

sebpop commented:
Mathieu, I just squashed the last patch in, and I removed the empty line.  Let me know if I need to change something else.
Thanks for your review!

mathieupoirier commented on tools/perf/util/cs-etm.c:
```diff
@@ -514,6 +538,65 @@ static int cs_etm__setup_queues(struct cs_etm_auxtrace *etm)
+
+        if (!etmq->last_branch_pos)
+                etmq->last_branch_pos = etmq->etm->synth_opts.last_branch_sz;
+
```
Add a comment that explains why ->last_branch_pos needs to be decremented by 1.

mathieupoirier commented on tools/perf/util/cs-etm.c:
```diff
@@ -514,6 +538,65 @@ static int cs_etm__setup_queues(struct cs_etm_auxtrace *etm)
+        be->flags.mispred = 0;
+        be->flags.predicted = 1;
+
+        if (bs->nr < etmq->etm->synth_opts.last_branch_sz)
```
Are we always absolutely sure the ->nr will be behind by only one?

sebpop replied:
bs->nr represents the number of branches already pushed to the bs (branch_stack).  It should be incremented until reaching the number of branches the user has specified in the command line "--itrace=il64", i.e., 64.  That number is made available in "etmq->etm->synth_opts.last_branch_sz".
Once bs->nr reaches that number it should remain set to that number: we then read that many entries from the branch stack and generate the last branch event.

mathieupoirier commented on tools/perf/util/cs-etm.c:
```diff
@@ -809,8 +911,13 @@ static int cs_etm__sample(struct cs_etm_queue *etmq)
+                return err;
+        if (etmq->etm->synth_opts.last_branch &&
+            etmq->last_branch_rb->nr < etmq->etm->synth_opts.last_branch_sz) {
+                // FIXME: need to synth partial stack on last event.
```
I this work you intend to do in this patchset?  What is currently missing that prevents you from implementing?

sebpop commented:
Mathieu,
I updated the patch and added more comments explaining how the circular buffers are managed.  I have removed the FIXME note as I implemented a way to drain the last branches from the circular buffer.
In reading the code again, I had to add two extra FIXME notes: one about trace sampling not working, and a question or assumption that I am making about the ETM traces, maybe you can help clearing that one out: does an ETM trace that we get from cs_etm__get_trace() only contain branches from a single cpu?

Thanks,
Sebastian

mathieupoirier commented:
On 14 December 2016 at 15:21, sebpop <notifications@github.com> wrote:

> Mathieu,
> I updated the patch and added more comments explaining how the circular
> buffers are managed. I have removed the FIXME note as I implemented a way
> to drain the last branches from the circular buffer.
> In reading the code again, I had to add two extra FIXME notes: one about
> trace sampling not working, and a question or assumption that I am making
> about the ETM traces, maybe you can help clearing that one out: does an ETM
> trace that we get from cs_etm__get_trace() only contain branches from a
> single cpu?
>
Good morning,

When adding FIXME notes, please add a comment that explains what is
currently blocking the implementation.

As for "cs_etm__get_trace(), I haven't authored the original code and it's
been a long time since I looked at it but it's taking input from the
perf.data file itself so I would say that it can contain traces from any
CPU.

Mathieu


> Thanks,
> Sebastian
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Linaro/OpenCSD/pull/8#issuecomment-267175524>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAi322yd-r2SypPlOWtCaJCOY57iXudjks5rIGv5gaJpZM4LNDz7>
> .
>

mikel-armbb commented:
This pull request has been applied to the latest perf-opencsd code - and further update.
Now in https://github.com/Linaro/perf-opencsd repository, autoFDO branch
Closing pull request as it does not apply here any more
